src/main/scala/edu/arizona/sista/odin/impl/ThompsonVM.scala

```diff
@@ -85,35 +95,35 @@ object ThompsonVM {
     // Instruction is expected to be a Match instruction.
     def stepSingleThread(t: SingleThread): Seq[Thread] = t.inst match {
       case i: MatchToken if t.tok < doc.sentences(sent).size && i.c.matches(t.tok, sent, doc, state) =>
-        mkThreads(t.tok + 1, i.next, t.groups, t.mentions, t.partialGroups)
+        val nextTok = if (t.dir == LeftToRight) t.tok + 1 else t.tok - 1
+        mkThreads(nextTok, i.next, t.dir, t.groups, t.mentions, t.partialGroups)
       case i: MatchSentenceStart if t.tok == 0 =>
-        mkThreads(t.tok, i.next, t.groups, t.mentions, t.partialGroups)
+        mkThreads(t.tok, i.next, t.dir, t.groups, t.mentions, t.partialGroups)
       case i: MatchSentenceEnd if t.tok == doc.sentences(sent).size =>
-        mkThreads(t.tok, i.next, t.groups, t.mentions, t.partialGroups)
+        mkThreads(t.tok, i.next, t.dir, t.groups, t.mentions, t.partialGroups)
       case i: MatchLookAhead =>
-        val results = eval(mkThreads(t.tok, i.start))
-        if (i.negative == results.isEmpty)
-          mkThreads(t.tok, i.next, t.groups, t.mentions, t.partialGroups)
-        else Nil
+        val results = eval(mkThreads(t.tok, i.start, LeftToRight))
+        if (i.negative == results.isEmpty) {
+          mkThreads(t.tok, i.next, t.dir, t.groups, t.mentions, t.partialGroups)
+        } else {
+          Nil
+        }
       case i: MatchLookBehind =>
-        val startTok = tok - i.size
-        if (startTok < 0) {
-          if (i.negative) mkThreads(t.tok, i.next, t.groups, t.mentions, t.partialGroups)
-          else Nil
+        val startTok = t.tok - 1
+        val results = if (startTok < 0) None else eval(mkThreads(startTok, i.start, RightToLeft))
+        if (i.negative == results.isEmpty) {
+          mkThreads(t.tok, i.next, t.dir, t.groups, t.mentions, t.partialGroups)
         } else {
-          val results = eval(mkThreads(startTok, i.start))
-          if (i.negative == results.isEmpty)
-            mkThreads(t.tok, i.next, t.groups, t.mentions, t.partialGroups)
-          else Nil
+          Nil
         }
       case i: MatchMention =>
         val bundles = for {
-          mention <- state.mentionsFor(sent, t.tok)
-          if mention.start == t.tok && mention.matches(i.m)
-        } yield {
-          val captures = mkMentionCapture(t.mentions, i.name, mention)
-          mkThreads(mention.end, i.next, t.groups, captures, t.partialGroups)
-        }
+          m <- state.mentionsFor(sent, t.tok)
+          if m.matches(i.m)
+          if (t.dir == LeftToRight && t.tok == m.start) || (t.dir == RightToLeft && t.tok == m.end - 1)
+          captures = mkMentionCapture(t.mentions, i.name, m)
+          nextTok = if (t.dir == LeftToRight) m.end else m.start - 1
```

Comment:
After matching a mention, the next token is determined based on the thread's direction